hail/src/main/java/is/hail/io/compress/BGzipInputStream.java

```diff
@@ -245,4 +253,23 @@ public void resetState() throws IOException {
             inputBufferPos = inputBufferSize;
         }
     }
+
+    // pos is a virtual file pointer, it is not a strict offset into the compressed data.
+    // The upper 48 bits of pos are the offset into the compressed data, the lower 16 bits
+    // are the offset into the uncompressed block that begins at the pointed to location
+    // by the upper 48 bits.
+    public void virtualSeek(final long pos) throws IOException {
+        final long compOff = BlockCompressedFilePointerUtil.getBlockAddress(pos);
+        final int uncompOff = BlockCompressedFilePointerUtil.getBlockOffset(pos);
+        if (currentBlockPos != compOff) {
+            ((Seekable) in).seek(compOff);
+            resetState(); // FIXME, this is probably less efficient than it could be, but it works for now
```

Comment:
So reset state searches forward to find the next start of block, but compOff needs to be a block start.  So I think you can change this to just update the input buffer state and call `decompressNextBlock`:

```
        inputBufferSize = 0;
        inputBufferPos = 0;
        inputBufferInPos = compOff;
        decompressNextBlock();
```